File: src/migrations/migration9.ts
import {Kingdom} from '../kingdom/data/kingdom';
import {Migration} from './migration';
import {recoverArmyIds} from '../kingdom/data/structures';

export class Migration9 extends Migration {
    constructor() {
        super(9);
    }

    override async migrateKingdom(game: Game, kingdom: Kingdom): Promise<void> {
        kingdom.modifiers.forEach(m => {
            const hasRecoverArmy = m.activities?.includes('recover-army');
            if (m.activities && hasRecoverArmy) {
                m.activities = [...m.activities.filter(a => a !== 'recover-army'), ...recoverArmyIds];
            }
        });
    }
}

File: src/jsMain/kotlin/at/posselt/pfrpg2e/migrations/migrations/Migration9.kt
package at.posselt.pfrpg2e.migrations.migrations

import at.posselt.pfrpg2e.kingdom.KingdomData
import com.foundryvtt.core.Game

class Migration9 : Migration(9) {
    override suspend fun migrateKingdom(game: Game, kingdom: KingdomData) {
        val recoverArmyIds = arrayOf(
            "recover-army-damaged",
            "recover-army-defeated",
            "recover-army-lost",
            "recover-army-mired-pinned",
            "recover-army-shaken",
            "recover-army-weary",
        )
        kingdom.modifiers.forEach { m ->
            val activities = m.asDynamic().activities
            if (activities != null && activities.contains("recover-army")) {
                m.asDynamic().activities = activities.unsafeCast<Array<String>>().filter { it != "recover-army" }.toTypedArray() + recoverArmyIds
            }
        }
    }
}